cirq/devices/grid_qubit.py

```diff
@@ -82,6 +64,197 @@ def neighbors(self,
                 neighbors.add(q)
         return neighbors
 
+    @abc.abstractmethod
+    def _with_row_col(self: TSelf, row: int, col: int) -> TSelf:
+        '''Returns a qid with the same type but a different coordinate.'''
+
+    def __add__(self: TSelf, other: Tuple[int, int]) -> 'TSelf':
+        if isinstance(other, _BaseGridQid):
+            return self._with_row_col(row=self.row + other.row, col=self.col + other.col)
+        if not (isinstance(other, tuple) and len(other) == 2 and
+                all(isinstance(x, int) for x in other)):
+            raise TypeError(
+                'Can only add integer tuples of length 2 to {}. Instead was {}'.format(
+                    type(self).__name__, other))
+        return self._with_col_row(row=self.row + other[0], col=self.col + other[1])
+
+    def __sub__(self: TSelf, other: Tuple[int, int]) -> 'TSelf':
+        if isinstance(other, GridQubit):
+            return self._with_col_row(row=self.row - other.row, col=self.col - other.col)
+        if not (isinstance(other, tuple) and len(other) == 2 and
+                all(isinstance(x, int) for x in other)):
+            raise TypeError(
+                'Can only subtract tuples of length 2 to GridQubits. Was {}'.
+                format(other))
+        return self._with_col_row(row=self.row - other[0], col=self.col - other[1])
+
+    def __radd__(self: TSelf, other: Tuple[int, int]) -> 'TSelf':
+        return self + other
+
+    def __rsub__(self, other: Tuple[int, int]) -> 'TSelf':
+        return -self + other
+
+    def __neg__(self: TSelf) -> 'TSelf':
+        return self._with_col_row(row=-self.row, col=-self.col)
+
+
+class GridQid(_BaseGridQid):
+    """A qid on a 2d square lattice
+
+    GridQid uses row-major ordering:
+
+        GridQid(0, 0, dimension=2) < GridQid(0, 1, dimension=2)
+        < GridQid(1, 0, dimension=2) < GridQid(1, 1, dimension=2)
+
+    New GridQid can be constructed by adding or subtracting tuples
+
+        >>> cirq.GridQubit(2, 3, dimension=2) + (3, 1)
+        cirq.GridQubit(5, 4, dimension=2)
+
+        >>> cirq.GridQubit(2, 3, dimension=2) - (1, 2)
+        cirq.GridQubit(1, 1, dimension=2)
+    """
+
+    def __init__(self, row: int, col: int, dimension: int) -> None:
```

Comment:
I think we should make dimension a forced keyword argument by putting `*` beforehand. This will avoid confusion where GridQid(3, 2, 4) could be seen as having dimension 3.